### PR TITLE
adding basic receiver vasp methods

### PR DESCRIPTION
**Pull request by matthappens**
# Receiver Vasp

This PR adds an endpoint to create an invoice, and an endpoint to create invoice + send to another uma address

# Testing
run a receiver vasp on port 8080, and a sender vasp on 8081

to create an invoice
`curl -u <username>:<password> -X POST "http://localhost:8080/api/uma/create_invoice?amount=100&currencyCode=USD"`

to create and send an invoice to another uma
`curl -u <username>:<password> -X POST "http://localhost:8080/api/uma/create_and_send_invoice/?amount=100&currencyCode=USD&senderUma=<sender uma>"`

### CODE REVIEW

**Comment by matthappens** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/lightsparkdev/kotlin-sdk/205?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#206** <a href="https://app.graphite.dev/github/pr/lightsparkdev/kotlin-sdk/206?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#205** <a href="https://app.graphite.dev/github/pr/lightsparkdev/kotlin-sdk/205?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#204** <a href="https://app.graphite.dev/github/pr/lightsparkdev/kotlin-sdk/204?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `develop`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @matthappens and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: umaserverdemo/src/main/kotlin/com/lightspark/ReceivingVasp.kt [hunk 1]
```diff
@@ -46,12 +61,130 @@ class ReceivingVasp(
...
+                "email" to false,
+                "compliance" to true,
+                "identifier" to true,
+            ),
+            callback = getLnurlpCallback(call), // structured the same, going to /api/uma/payreq/{user_id}
+            privateSigningKey = config.umaSigningPrivKey,
+            senderUma = senderUma
+        )
+        val encodedInvoice = invoice.toBech32()
+        val response = httpClient.post("/api/uma/payreq/${config.userID}") {
```
Comment (matthappens) [inline umaserverdemo/src/main/kotlin/com/lightspark/ReceivingVasp.kt:179]: add try catch

File: umaserverdemo/src/main/kotlin/com/lightspark/ReceivingVasp.kt [hunk 2]
```diff
@@ -46,12 +61,130 @@ class ReceivingVasp(
...
+                return "Amount not provided."
+            }
+        } catch (e: NumberFormatException) {
+            call.respond(HttpStatusCode.BadRequest, "Amount not parsable as number.")
+            return "Amount not parsable as number."
+        }
+
+        val currency = call.parameters["currencyCode"]?.let { currencyCode ->
+            // check if we support this currency code.
+            getReceivingCurrencies(UMA_VERSION_STRING).firstOrNull {
```
Comment (matthappens) [inline umaserverdemo/src/main/kotlin/com/lightspark/ReceivingVasp.kt:143]: much of this function duplicates invoice creation , should consolidate

File: umaserverdemo/src/main/kotlin/com/lightspark/ReceivingVasp.kt [hunk 3]
```diff
@@ -46,15 +60,103 @@ class ReceivingVasp(
...
+            call.respond(HttpStatusCode.BadRequest, "SenderUma not provided.")
+            return "SenderUma not provided."
+        }
+        val (status, data) = createUmaInvoice(call, senderUma)
+        if (status != HttpStatusCode.OK) {
+            call.respond(status, data)
+            return data
+        }
+        val response = try {
+            httpClient.post("/api/uma/payreq/${config.userID}") {
```
Comment (jklein24) [inline umaserverdemo/src/main/kotlin/com/lightspark/ReceivingVasp.kt:95]: I'm confused why we're sending to the payreq endpoint. Also what host is this hitting?

Comment (matthappens) [reply umaserverdemo/src/main/kotlin/com/lightspark/ReceivingVasp.kt:95]: yeah this one is incorrect I realized when testing - it should be : /api/uma/request_invoice_payment

Comment (matthappens) [reply umaserverdemo/src/main/kotlin/com/lightspark/ReceivingVasp.kt:95]: additionally - it should use the domain of the sender, which I have to adjust

Comment (jklein24) [reply umaserverdemo/src/main/kotlin/com/lightspark/ReceivingVasp.kt:95]: Shouldn't it be reading the route from the configuration discovery document?

Comment (matthappens) [reply umaserverdemo/src/main/kotlin/com/lightspark/ReceivingVasp.kt]: done

File: umaserverdemo/src/main/kotlin/com/lightspark/ReceivingVasp.kt [hunk 4]
```diff
@@ -46,15 +64,128 @@ class ReceivingVasp(
...
+            val response = httpClient.get(wellKnownConfiguration)
+            if (response.status != HttpStatusCode.OK) {
+                call.respond(
+                    HttpStatusCode.FailedDependency,
+                    "failed to fetch request / pay endpoint at $wellKnownConfiguration"
+                )
+                return "failed to fetch request / pay endpoint at $wellKnownConfiguration"
+            } else {
+                Json.decodeFromString<JsonObject>(
+                    response.bodyAsText())["uma_request_endpoint"]?.jsonPrimitive?.content ?: ""
```
Comment (jklein24) [inline umaserverdemo/src/main/kotlin/com/lightspark/ReceivingVasp.kt:114]: If the uma_request_endpoint is empty (in the `?:` case), we should fail, right?

Comment (matthappens) [reply umaserverdemo/src/main/kotlin/com/lightspark/ReceivingVasp.kt:114]: agreed

**Review by jklein24 — APPROVED** [review]

**Pull request merged**